katpoint/target.py

```diff
@@ -618,24 +618,42 @@ def uvw_basis(self, timestamp=None, antenna=None):
             dimension to the timestamp.
         """
         timestamp, antenna = self._set_timestamp_antenna_defaults(timestamp, antenna)
-        # NCP vector is J2000 NCP
-        ncp = construct_radec_target(0.0, np.pi / 2.0)
-        # Get J2000 NCP az-el vector at current epoch pointed to by reference antenna
-        ncp_az, ncp_el = ncp.azel(timestamp, antenna)
+        if is_iterable(timestamp) and self.body_type != 'radec':
+            # Some calculations depend on ra/dec in a way that won't easily
+            # vectorise.
+            bases = [self.uvw_basis(t, antenna) for t in timestamp]
+            return np.stack(bases, axis=-1)
+
+        # Offset the target slightly in declination to approximate the
+        # derivative of ENU in the direction of increasing declination. This
+        # used to just use the NCP, but the astrometric-to-topocentric
+        # conversion doesn't simply rotate the celestial sphere, but also
+        # distorts it, and so that introduced errors.
+        #
+        # To avoid issues close to the poles, we always offset towards the
+        # equator. We also can't offset by too little, as ephem uses only
+        # single precision and this method suffers from loss of precision.
+        # 0.03 was found by experimentation (albeit on a single data set) to
```

Comment:
0.03 radians, I assume. That's about 2 degrees... Interesting that one can go that big. I guess it is a mild distortion to begin with.

Reply:
I suspect we'd get slightly better results if we could go smaller, but then the floating point errors climb rapidly.